integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -562,18 +567,28 @@ def setUpClass(cls) -> None:
     # Setup the directory for saving compilation artifacts and traces.
     cls._artifacts_dir = _setup_artifacts_dir(cls._module_class.__name__)
 
-  def setUp(self):
-    # Ran before each unit test.
+    # Get the backend information for this test.
+    cls._ref_backend_info = tf_utils.BackendInfo(
+        FLAGS.reference_backend, f"{FLAGS.reference_backend}_ref")
+    cls._tar_backend_infos = get_target_backends()
+
+  def setUp(self) -> None:
+    # Runs before each unit test.
     super().setUp()
-    # Create a CompiledModule for the reference backend and each target backend.
-    ref_backend_info = tf_utils.BackendInfo(FLAGS.reference_backend,
-                                            f"{FLAGS.reference_backend}_ref")
-    self._ref_module = self._compile(ref_backend_info)
-
-    tar_backend_infos = get_target_backends()
-    self._tar_modules = [
-        self._compile(backend_info) for backend_info in tar_backend_infos
-    ]
+    global _global_ref_module
+    global _global_tar_modules
+    if _global_ref_module is None:
+      # Compile if this is the first unittest.
+      _global_ref_module = self._compile(self._ref_backend_info)
+      _global_tar_modules = [
+          self._compile(backend_info)
+          for backend_info in self._tar_backend_infos
+      ]
+    else:
+      _global_ref_module = _global_ref_module.create_reinitialized()
+      _global_tar_modules = [
+          module.create_reinitialized() for module in _global_tar_modules
```

Comment:
Another thing to consider is whether we can have a version of modules that stores their key compiled information without being the thing responsible for holding onto context and such. Then that could be something stored on the class and we would construct a stateful module that wraps that and does context setup and such. I think that would be cleaner than bundling compilation in with everything else. It would also avoid this "create reinitialized" thing, which honestly looks pretty weird to me... Would be more work though, so can certainly be a later change.

Reply:
I could write something like this:

```python
class IreeCompiledModuleInfo:

  def __init__(self, module_class, backend_info, exported_names, artifacts_dir=None):
    self._module_class = module_class
    ...

    set_random_seed()
    self._module_blob = compile_tf_module(
        tf_module=module_class(),
        backend_infos=[backend_info],
        exported_names=exported_names,
        artifacts_dir=artifacts_dir)
    self._module = rt.VmModule.from_flatbuffer(self._module_blob)

  def instantiate(self):
    ...
    return IreeCompiledModule(self._module_class, ..., _secret_squirrel_dict)


class IreeCompiledModule(CompiledModule):

  def __init__(self, 
               module_class,
               backend_info,
               exported_names, 
               artifacts_dir=None, 
               _secret_squirrel_dict=None):
    super(...)
    
    if _secret_squirrel_args is None:
      # Compile and get VmModule
      self._module = ...
    else:
      self._module = _secret_squirrel_dict["_module"]
    
    self._context = ...
```

This would allow us to put the `IreeCompiledModuleInfo` on the `cls` and then create `IreeCompiledModule`s via `instantiate` for the test suite, but still provide a non-convoluted API for just creating an `IreeCompiledModule` if you want to.

Reply:
Seems like it's easy to provide a function for compiling and then initializing if we have those things separate, but hard to break things out if you don't. I would rather the init be simple and do no initialization. One option would be to still have one class, but have a separate `initialize` function that has to be called before it's used and then a `createInitialized` function that does both. `initialize` functions are usually a bad API though because then anytime you do anything with it you have to make sure it's initialized first. Hence just having a separate class for it.

Reply:
Table this here and dedupe the convo?